### PR TITLE
SpectrumInfo and DetectorInfo rollout: algorithms <LoadMaskTest.h to SaveParameterFileTest.h>

### LINKED ISSUE #18578: SpectrumInfo and DetectorInfo rollout: algorithms <LoadMaskTest.h to SaveParameterFileTest.h>
This issue is part of the SpectrumInfo and DetectorInfo rollout, see #17743.
Algorithms covered in this issue are:
```
Framework/DataHandling/.../LoadMaskTest.h
Framework/DataHandling/.../LoadGSSTest.h
Framework/DataHandling/.../LoadMuonNexus2Test.h
Framework/DataHandling/.../LoadParameterFileTest.h
Framework/DataHandling/.../LoadQKKTest.h
Framework/DataHandling/.../LoadRaw3Test.h
Framework/DataHandling/.../LoadRawSaveNxsLoadNxsTest.h
Framework/DataHandling/.../LoadSpiceXML2DDetTest.h
Framework/DataHandling/.../MaskDetectorsTest.h
Framework/DataHandling/.../SaveParameterFileTest.h
```

### PR DESCRIPTION
**Pull request by AntonPiccardoSelg**
## Description of work.
This PR is part of refactoring effort to use the new interface of the SpectrumInfo and DetectorInfo modules.
See https://github.com/mantidproject/mantid/issues/17743 for details.

The following algorithms and tests were refactored to use SpectrumInfo/DetectorInfo:
```
Framework/DataHandling/.../LoadMaskTest.h
Framework/DataHandling/.../LoadGSSTest.h
Framework/DataHandling/.../LoadMuonNexus2Test.h
Framework/DataHandling/.../LoadParameterFileTest.h
Framework/DataHandling/.../LoadQKKTest.h
Framework/DataHandling/.../LoadRaw3Test.h
Framework/DataHandling/.../LoadRawSaveNxsLoadNxsTest.h
Framework/DataHandling/.../LoadSpiceXML2DDetTest.h
Framework/DataHandling/.../MaskDetectorsTest.h
Framework/DataHandling/.../SaveParameterFileTest.h
```

Performance tests were added for:
Most changes are in test files and unrelated to performance tests.

The following algorithms have improved performance:
N/A

## To test.

Code review.

Fixes #18578
Internal change, no release notes (potential performance improvements will be added to release notes as part of the umbrella issue).

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/DataHandling/test/LoadGSSTest.h [hunk 1]
```diff
@@ -89,14 +90,15 @@ class LoadGSSTest : public CxxTest::TestSuite {
     const auto sample = ws->getInstrument()->getSample();
     TS_ASSERT_DELTA(sample->getDistance(*source), 40., 1e-4);
 
-    const auto det0 = ws->getDetector(0);
-    TS_ASSERT_DELTA(det0->getDistance(*sample), 2.2222, 1e-4);
-    TS_ASSERT_DELTA(det0->getTwoTheta(V3D(0.0, 0.0, 0.0), V3D(0.0, 0.0, 1.0)) *
+    const auto &spectrumInfo = ws->spectrumInfo();
+    const auto &det0 = spectrumInfo.detector(0);
+    TS_ASSERT_DELTA(det0.getDistance(*sample), 2.2222, 1e-4);
```
Comment (SimonHeybrock) [inline Framework/DataHandling/test/LoadGSSTest.h]: `SpectrumInfo::l2`?

File: Framework/DataHandling/test/LoadGSSTest.h [hunk 2]
```diff
@@ -89,14 +90,15 @@ class LoadGSSTest : public CxxTest::TestSuite {
...
+    TS_ASSERT_DELTA(det0.getTwoTheta(V3D(0.0, 0.0, 0.0), V3D(0.0, 0.0, 1.0)) *
```
Comment (SimonHeybrock) [inline Framework/DataHandling/test/LoadGSSTest.h]: `SpectrumInfo::twoTheta`?

Comment (SimonHeybrock) [inline Framework/DataHandling/test/LoadGSSTest.h]: (some below)

File: Framework/DataHandling/test/LoadQKKTest.h [hunk 3]
```diff
@@ -50,9 +51,9 @@ class LoadQKKTest : public CxxTest::TestSuite {
     Workspace2D_sptr data = boost::dynamic_pointer_cast<Workspace2D>(ws);
     TS_ASSERT(data);
     TS_ASSERT_EQUALS(data->getNumberHistograms(), 192 * 192);
-    Mantid::Geometry::IDetector_const_sptr det;
+    const auto &spectrumInfo = data->spectrumInfo();
     for (size_t i = 0; i < data->getNumberHistograms(); ++i) {
-      TS_ASSERT_THROWS_NOTHING(det = data->getDetector(i));
+      TS_ASSERT_THROWS_NOTHING(const auto &det = spectrumInfo.detector(i));
```
Comment (SimonHeybrock) [inline Framework/DataHandling/test/LoadQKKTest.h]: Maybe use `TS_ASSERT(spectrumInfo.hasDetectors(i))`

File: Framework/DataHandling/test/LoadRawSaveNxsLoadNxsTest.h [hunk 4]
```diff
@@ -156,16 +157,25 @@ class LoadRawSaveNxsLoadNxsTest : public CxxTest::TestSuite {
       TS_ASSERT_EQUALS(samplepos->getName(), "some-surface-holder");
       TS_ASSERT_DELTA(samplepos->getPos().X(), 0.0, 0.01);
```
Comment (SimonHeybrock) [inline Framework/DataHandling/test/LoadRawSaveNxsLoadNxsTest.h]: There is `DetectorInfo::sourcePosition()` and `samplePosition()`.

File: Framework/DataHandling/test/LoadSpiceXML2DDetTest.h [hunk 5]
```diff
@@ -177,7 +178,8 @@ class LoadSpiceXML2DDetTest : public CxxTest::TestSuite {
     size_t center_col = 128;
     size_t center_row = 115;
     size_t center_ws_index = (center_row - 1) + (center_col - 1) * 256;
-    Kernel::V3D det_center = outws->getDetector(center_ws_index)->getPos();
+    const auto &spectrumInfo = outws->spectrumInfo();
+    const auto det_center = spectrumInfo.position(center_ws_index);
     // distance to sample
     double dist_r = det_center.distance(sample);
```
Comment (SimonHeybrock) [inline Framework/DataHandling/test/LoadSpiceXML2DDetTest.h]: `spectrumInfo.l2`.

File: Framework/DataHandling/test/LoadSpiceXML2DDetTest.h [hunk 6]
```diff
@@ -336,7 +342,8 @@ class LoadSpiceXML2DDetTest : public CxxTest::TestSuite {
     size_t center_row = 115 - 1;
     size_t center_col = 128 - 1;
     size_t center_ws_index = 256 * center_col + center_row;
-    Kernel::V3D center_det_pos = outws->getDetector(center_ws_index)->getPos();
+    const auto &spectrumInfo = outws->spectrumInfo();
+    const auto center_det_pos = spectrumInfo.position(center_ws_index);
     TS_ASSERT_DELTA(center_det_pos.Y(), 0., 0.00000001);
     double sample_center_distance = sample.distance(center_det_pos);
```
Comment (SimonHeybrock) [inline Framework/DataHandling/test/LoadSpiceXML2DDetTest.h]: `SpectrumInfo::l2`

File: Framework/DataHandling/test/LoadSpiceXML2DDetTest.h [hunk 7]
```diff
@@ -346,12 +353,12 @@ class LoadSpiceXML2DDetTest : public CxxTest::TestSuite {
     TS_ASSERT_DELTA(sample_center_angle * 180. / M_PI, twotheta_log, 0.0001);
 
     size_t ll_ws_index = 0;
-    Kernel::V3D ll_det_pos = outws->getDetector(ll_ws_index)->getPos();
+    const auto ll_det_pos = spectrumInfo.position(ll_ws_index);
     double ll_sample_r = sample.distance(ll_det_pos);
```
Comment (SimonHeybrock) [inline Framework/DataHandling/test/LoadSpiceXML2DDetTest.h]: `l2`

File: Framework/DataHandling/test/LoadSpiceXML2DDetTest.h [hunk 8]
```diff
@@ -346,12 +353,12 @@ class LoadSpiceXML2DDetTest : public CxxTest::TestSuite {
...
     TS_ASSERT_DELTA(ll_sample_r, 0.37597, 0.001);
 
     size_t lu_ws_index = 255; // row = 255, col = 1
-    Kernel::V3D lu_det_pos = outws->getDetector(lu_ws_index)->getPos();
+    const auto lu_det_pos = spectrumInfo.position(lu_ws_index);
     double lu_sample_r = sample.distance(lu_det_pos);
```
Comment (SimonHeybrock) [inline Framework/DataHandling/test/LoadSpiceXML2DDetTest.h]: `l2`

File: Framework/DataHandling/test/LoadSpiceXML2DDetTest.h [hunk 9]
```diff
@@ -417,7 +424,8 @@ class LoadSpiceXML2DDetTest : public CxxTest::TestSuite {
     size_t center_col = 128;
     size_t center_row = 115;
     size_t center_ws_index = (center_row - 1) + (center_col - 1) * 256;
-    Kernel::V3D center_det_pos = outws->getDetector(center_ws_index)->getPos();
+    const auto &spectrumInfo = outws->spectrumInfo();
+    const auto center_det_pos = spectrumInfo.position(center_ws_index);
     // distance to sample
     double dist_r = center_det_pos.distance(sample);
```
Comment (SimonHeybrock) [inline Framework/DataHandling/test/LoadSpiceXML2DDetTest.h]: `l2`

File: Framework/DataHandling/test/LoadSpiceXML2DDetTest.h [hunk 10]
```diff
@@ -542,7 +550,9 @@ class LoadSpiceXML2DDetTest : public CxxTest::TestSuite {
     size_t center_col = 137 - 1;
     size_t center_ws_index = 256 * center_col + center_row;
     // y should be 0. in the Z-Y plane
-    Kernel::V3D center_det_pos = outws->getDetector(center_ws_index)->getPos();
+    const auto &spectrumInfo = outws->spectrumInfo();
+    const auto center_det_pos = spectrumInfo.position(center_ws_index);
+
     TS_ASSERT_DELTA(center_det_pos.Y(), 0., 0.00000001);
     double sample_center_distance = sample.distance(center_det_pos);
```
Comment (SimonHeybrock) [inline Framework/DataHandling/test/LoadSpiceXML2DDetTest.h]: `l2`

File: Framework/DataHandling/test/LoadSpiceXML2DDetTest.h [hunk 11]
```diff
@@ -559,26 +569,29 @@ class LoadSpiceXML2DDetTest : public CxxTest::TestSuite {
 
     size_t ll_ws_index =
         (center_row - ws_d_row) + (center_col - ws_d_col) * 256;
-    Kernel::V3D ll_det_pos = outws->getDetector(ll_ws_index)->getPos();
+    const auto ll_det_pos = spectrumInfo.position(ll_ws_index);
+
     double ll_sample_r = sample.distance(ll_det_pos);
```
Comment (SimonHeybrock) [inline Framework/DataHandling/test/LoadSpiceXML2DDetTest.h]: `l2`

File: Framework/DataHandling/test/LoadSpiceXML2DDetTest.h [hunk 12]
```diff
@@ -559,26 +569,29 @@ class LoadSpiceXML2DDetTest : public CxxTest::TestSuite {
...
 
     size_t lr_ws_index =
         (center_row + ws_d_row) + (center_col - ws_d_col) * 256;
-    Kernel::V3D lr_det_pos = outws->getDetector(lr_ws_index)->getPos();
+    const auto lr_det_pos = spectrumInfo.position(lr_ws_index);
+
     double lr_sample_r = sample.distance(lr_det_pos);
```
Comment (SimonHeybrock) [inline Framework/DataHandling/test/LoadSpiceXML2DDetTest.h]: `l2`

File: Framework/DataHandling/test/LoadSpiceXML2DDetTest.h [hunk 13]
```diff
@@ -559,26 +569,29 @@ class LoadSpiceXML2DDetTest : public CxxTest::TestSuite {
...
 
     TS_ASSERT_DELTA(ll_sample_r, lr_sample_r, 0.0000001);
 
     size_t ur_ws_index =
         (center_row + ws_d_row) + (center_col + ws_d_col) * 256;
-    Kernel::V3D ur_det_pos = outws->getDetector(ur_ws_index)->getPos();
+    const auto ur_det_pos = spectrumInfo.position(ur_ws_index);
+
     double ur_sample_r = sample.distance(ur_det_pos);
```
Comment (SimonHeybrock) [inline Framework/DataHandling/test/LoadSpiceXML2DDetTest.h]: `l2`

File: Framework/DataHandling/test/LoadSpiceXML2DDetTest.h [hunk 14]
```diff
@@ -559,26 +569,29 @@ class LoadSpiceXML2DDetTest : public CxxTest::TestSuite {
...
 
     TS_ASSERT_DELTA(ll_sample_r, ur_sample_r, 0.0000001);
 
     size_t ul_ws_index =
         (center_row - ws_d_row) + (center_col + ws_d_col) * 256;
-    Kernel::V3D ul_det_pos = outws->getDetector(ul_ws_index)->getPos();
+    const auto ul_det_pos = spectrumInfo.position(ul_ws_index);
     double ul_sample_r = sample.distance(ul_det_pos);
```
Comment (SimonHeybrock) [inline Framework/DataHandling/test/LoadSpiceXML2DDetTest.h]: `l2`

**Review by SimonHeybrock — COMMENTED** [review]
Some comments below:

File: Framework/DataHandling/test/LoadSpiceXML2DDetTest.h [hunk 15]
```diff
@@ -335,17 +335,17 @@ class LoadSpiceXML2DDetTest : public CxxTest::TestSuite {
...
     size_t center_col = 128 - 1;
     size_t center_ws_index = 256 * center_col + center_row;
-    const auto &spectrumInfo = outws->spectrumInfo();
     const auto center_det_pos = spectrumInfo.position(center_ws_index);
     TS_ASSERT_DELTA(center_det_pos.Y(), 0., 0.00000001);
-    double sample_center_distance = sample.distance(center_det_pos);
+    double sample_center_distance = spectrumInfo.l2(center_ws_index);
     // TS_ASSERT_DELTA(center_det_pos.X(), )
     TS_ASSERT_DELTA(sample_center_distance, 0.3750, 0.0000001);
     double sample_center_angle =
```
Comment (SimonHeybrock) [inline Framework/DataHandling/test/LoadSpiceXML2DDetTest.h]: `SpectrumInfo::twoTheta`?

File: Framework/DataHandling/test/LoadRawSaveNxsLoadNxsTest.h [hunk 16]
```diff
@@ -145,27 +146,37 @@ class LoadRawSaveNxsLoadNxsTest : public CxxTest::TestSuite {
...
+      }
+
+      if (hasDetector) {
+        const auto &detector103 = detectorInfo.detector(detectorIndex);
+        TS_ASSERT_EQUALS(detector103.getID(), 103);
+        TS_ASSERT_EQUALS(detector103.getName(), "linear-detector-pixel");
+        TS_ASSERT_DELTA(detectorInfo.position(detectorIndex).Z(), 12.403, 0.01);
+        TS_ASSERT_DELTA(detectorInfo.position(detectorIndex).Y(), 0.1164, 0.01);
+        double d =
+            detectorInfo.position(detectorIndex).distance(samplepos->getPos());
```
Comment (SimonHeybrock) [inline Framework/DataHandling/test/LoadRawSaveNxsLoadNxsTest.h]: `DetectorInfo::l2`?

**Review by SimonHeybrock — COMMENTED** [review]
Update looks good, just two more minor improvements:

**Comment by SimonHeybrock** [comment]
Code looks good now, :shipit:

**Comment by ianbush** [comment]
@AntonPiccardoSelg Why was `SaveParameterFileTest.h` reverted?

**Comment by AntonPiccardoSelg** [comment]
@ianbush  You are right, I must have forgotten to re-revert it. It had been causing some serious pain. 

Essentially the combination of `DetectorInfo` and `ws->instrumentParameters()` is dangerous. The call to `instrumentParameters()` invalidates some members of a preceding `const auto& detectorInfo = ws->detectorInfo()` instance. Unless it is the aim to get rid of `instrumentParameters` it might make sense to advertise this danger or to hedge  it.

**Comment by ianbush** [comment]
@AntonPiccardoSelg Yes, you are right we should advertise this better. Eventually the parameter map should be used a lot less.

**Pull request merged**